server/src/internalClusterTest/java/org/elasticsearch/rest/ChunkedZipResponseIT.java

```diff
@@ -0,0 +1,437 @@
+        public boolean isPartComplete() {
+            return chunksIterator.hasNext() == false;
+        }
+
+        @Override
+        public boolean isLastPart() {
+            return partsIterator.hasNext() == false;
+        }
+
+        @Override
+        public void getNextPart(ActionListener<ChunkedRestResponseBodyPart> listener) {
+            threadPool.generic()
+                .execute(ActionRunnable.supply(listener, () -> new TestBytesReferenceBodyPart(name, threadPool, partsIterator, refs)));
+        }
+
+        @Override
+        public ReleasableBytesReference encodeChunk(int sizeHint, Recycler<BytesRef> recycler) {
+            assert chunksIterator.hasNext();
+            return new ReleasableBytesReference(chunksIterator.next(), refs.acquire());
+        }
+
+        @Override
+        public String getResponseContentTypeString() {
+            return "application/binary";
+        }
+    }
+
+    public void testRandomZipResponse() throws IOException {
+        final var request = new Request("GET", RandomZipResponsePlugin.ROUTE);
+        if (randomBoolean()) {
+            request.setOptions(
+                RequestOptions.DEFAULT.toBuilder()
+                    .addHeader("accept-encoding", String.join(", ", randomSubsetOf(List.of("deflate", "gzip", "zstd", "br"))))
+            );
+        }
+        final var response = getRestClient().performRequest(request);
+        assertEquals(ZIP_CONTENT_TYPE, response.getHeader("Content-Type"));
+        assertNull(response.getHeader("content-encoding")); // zip file is already compressed
+        assertEquals(
+            "attachment; filename=\"" + RandomZipResponsePlugin.RESPONSE_FILENAME + ".zip\"",
+            response.getHeader("Content-Disposition")
+        );
+        final var pathPrefix = RandomZipResponsePlugin.RESPONSE_FILENAME + "/";
+
+        final var actualEntries = new HashMap<String, BytesReference>();
+        final var copyBuffer = new byte[PageCacheRecycler.BYTE_PAGE_SIZE];
+
+        try (var zipStream = new ZipInputStream(response.getEntity().getContent())) {
+            ZipEntry zipEntry;
+            while ((zipEntry = zipStream.getNextEntry()) != null) {
+                assertThat(zipEntry.getName(), startsWith(pathPrefix));
+                final var name = zipEntry.getName().substring(pathPrefix.length());
+                try (var bytesStream = new BytesStreamOutput()) {
+                    while (true) {
+                        final var readLength = zipStream.read(copyBuffer, 0, copyBuffer.length);
+                        if (readLength < 0) {
+                            break;
+                        }
+                        bytesStream.write(copyBuffer, 0, readLength);
+                    }
+                    actualEntries.put(name, bytesStream.bytes());
+                }
+            }
+        }
+
+        assertEquals(getExpectedEntries(), actualEntries);
+    }
+
+    public void testAbort() throws IOException {
```

Comment:
The response can also be closed from the server side. Should we have a test for that?

Reply:
Ah wait I see, do you mean we should also test the handling of a `getNextPart()` failure? Sure, let me try and do that.

Reply:
Ok I added a test covering both a `getNextPart()` failure and a server-side channel close in  c2812c0c621223a62ad16a4eb9dfe5713d47d801.